src/lib/shadow-shim-helper-rs/src/rootedcell/rc.rs

```diff
@@ -0,0 +1,259 @@
+/// or leaking the object in release builds.
+pub struct RootedRc<T> {
+    tag: Tag,
+    internal: *mut RootedRcInternal<T>,
+}
+
+impl<T> RootedRc<T> {
+    /// Creates a new object associated with `root`.
+    pub fn new(root: &Root, val: T) -> Self {
+        Self {
+            tag: root.tag(),
+            internal: Box::into_raw(Box::new(RootedRcInternal::new(val))),
+        }
+    }
+
+    /// Like [Clone::clone], but requires that the corresponding Root is held.
+    ///
+    /// Intentionally named clone to shadow Self::deref()::clone().
+    ///
+    /// Panics if `root` is not the associated [Root].
+    pub fn clone(&self, root: &Root) -> Self {
+        assert_eq!(
+            root.tag, self.tag,
+            "Tried using root {:?} instead of {:?}",
+            root.tag, self.tag
+        );
+        // SAFETY: We've verified that the lock is held by inspection of the
+        // lock itself. We hold a reference to the guard, guaranteeing that the
+        // lock is held while `unchecked_clone` runs.
+        unsafe { self.unchecked_clone() }
+    }
+
+    /// # Safety
+    ///
+    /// There must be no other threads accessing this object, or clones of this object.
+    unsafe fn unchecked_clone(&self) -> Self {
+        // SAFETY: Pointer should be valid by construction. Caller is
+        // responsible for ensuring no parallel access.
+        let internal = unsafe { self.internal.as_ref().unwrap() };
+        internal.inc_strong();
+        Self {
+            tag: self.tag,
+            internal: self.internal,
+        }
+    }
+
+    /// Safely drop this object, dropping the internal value if no other
+    /// references to it remain.
+    ///
+    /// Instances that are dropped *without* calling this method cannot be
+    /// safely cleaned up. In debug builds this will result in a `panic`.
+    /// Otherwise the underlying reference count will simply not be decremented,
+    /// ultimately resulting in the enclosed value never being dropped.
+    pub fn safely_drop(mut self, root: &Root) {
+        assert_eq!(
+            root.tag, self.tag,
+            "Tried using a root {:?} instead of {:?}",
+            root.tag, self.tag
+        );
+        let drop_internal = {
+            // SAFETY: pointer points to valid data by construction.
+            let internal = unsafe { self.internal.as_ref() }.unwrap();
+            internal.dec_strong();
+            internal.strong_count.get() == 0
+        };
+        if drop_internal {
+            // SAFETY: There are no remaining strong references to
+            // self.internal, and we know that no other threads could be
+            // manipulating the reference count in parallel since we have the
+            // root lock.
+            unsafe { Box::from_raw(self.internal) };
+        }
+        self.internal = std::ptr::null_mut();
+    }
+}
+
+impl<T> Drop for RootedRc<T> {
+    fn drop(&mut self) {
```

Comment:
You might be able to do something like the following:

```rust
#[inline]
fn drop(&mut self) {
    struct C<T>(T);
    impl<T> C<T> {
        const V: core::convert::Infallible = panic!("you shouldn't drop RootedRc!");
    }
    match C::<T>::V {}
}
```

(adapted from https://medium.com/@adetaylor/are-we-reference-yet-c-references-in-rust-72c1c6c7015a)

Since the "inline" is just a compiler suggestion, I'm not sure if it can be relied on, but may be useful for catching at least some accidental drops.